File: wordgame.py
#!/bin/python
import sys

from graph import *

def findWeight(vertex1, vertex2):
	""" Finds the weight between two vertices in accordance to specs given in homework """

	# assume lengths will be the same
	numDiff = 0
	word1 = vertex1.word
	word2 = vertex2.word
	for n in range(0, len(vertex1.word)):
		if word1[n] != word2[n]:
			numDiff += 1

		if numDiff > 2:
			return None

	# weight the edges 
	if numDiff == 1:
		return 1
	elif numDiff == 2:
		return 5
	else:
		return 0

if __name__ == "__main__":
	# make sure there are the proper number of arguments
	if len(sys.argv) != 2:
		print("Word game program usage:")
		print("wordgame.py [filename]")
		exit()

	# get the file contents and close immediately
	filename = sys.argv[1]
	fh = open(filename, "r")
	fileContent = fh.read()
	fh.close()

	# build the graph #

	# add all the vertices
	graph = Graph()
	for word in fileContent.split():
		# create a new vertex and add it to the graph
		vertex = Vertex(word)
		graph.addVertex(vertex)

	# compare every combination of vertices
	for n in range(0, len(graph.vertices)):
		for m in range(n, len(graph.vertices)):
			weight = findWeight(graph.vertices[n], graph.vertices[m])

			# add the edge if it is valid
			if weight == 1 or weight == 5:
				graph.vertices[n].connect(graph.vertices[m], weight)

	# main program loop
	while True:
		# propt user for a five-letter word and convert to upper case
		inputStr = input("Enter a five-letter word: ")
		inputStr = inputStr.upper()

		# get the vertex from graph
		vertex = graph.getVertexByWord(inputStr)
		if vertex is None:
			print("The word you entered doesn't exist in the graph, please try again")
			continue

		print("The neighbors of " + inputStr + " are :")
		counter = 1
		for n in range(0, len(vertex.adj)):
			adj = vertex.adj[n]
			weight = vertex.weights[n]
			
			end = ""
			if counter == 6:
				end = "\n"
				counter = 0

			# print out the adjacent element
			print("\t" + adj.word + " (" + str(weight) + ")", end=end)
			counter += 1

		# add a newline at the end
		print()

File: src/WordGame.java
import java.io.BufferedReader;
import java.io.Console;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.PriorityQueue;
import java.util.Set;

public class WordGame {
	public static int findWeight(Vertex v1, Vertex v2) {
		int numDiff = 0;
		String word1 = v1.getWord();
		String word2 = v2.getWord();
		
		// assume words are both the same length
		for (int n = 0; n < word1.length(); n++) {
			char c1 = word1.charAt(n);
			char c2 = word2.charAt(n);
			
			if(c1 != c2){
				numDiff += 1;
			}
			if(numDiff > 2){
				return 0;
			}
		}
		
		// compute the weight from the word differences
		if(numDiff == 1){
			return 1;
		}
		else if (numDiff == 2){
			return 5;
		}
		else{
			return 0;
		}
	}
	
	public static String readWord(FileReader reader) throws IOException {
		String word = "";
		
		while(true) {
			// get the next character
			int byteRead = reader.read();
			if (byteRead < 0) {
				return "eof";
			}
			
			char c = (char)byteRead;
			if (c == ' ' || c == '\n') {
				break;
			}
			
			// add the character to the word
			word += c;
		}
		
		// don't return a word if the length is not 5
		if (word.length() != 5) {
			return null;
		}
		
		return word;
	}
	
	public void buildDijstra(Graph graph, String startWord) {
		PriorityQueue<Vertex> queue = new PriorityQueue<Vertex>();
		
		// the start vertex has a distance of zero
		Vertex startVertex = graph.getVertex(startWord);
		startVertex.setDistance(0);
		
		// put the first vertex in the queue
		queue.add(startVertex);
		
		// loop until the queue is empty
		while (!queue.isEmpty()) {
			// get the next vertex
			Vertex vertex = queue.poll();
			
			// iterate through the adjacency list
			Vertex[] adjList = vertex.getAdjList();
			for (Vertex adjVertex : adjList) {

				// compute the distance to the adjacent vertex using this shortest path
				int vertexDist = vertex.getDistance();
				int edgeWeight = vertex.getWeight(adjVertex);
				int adjVertexDist = vertexDist + edgeWeight;
				
				// set the new distance if the current distance is greater than the newly calculated one
				if (adjVertex.getDistance() > adjVertexDist) {
					
					// set that vertex's distance
					adjVertex.setDistance(adjVertexDist);
					
					// add that vertex to the queue
					queue.add(adjVertex);
				}
			}
		}
	}
	
	public static void main(String[] args) throws IOException {
		// make sure there are proper number of arguments
		if(args.length != 1){
			System.out.println("Word game program usage:");
			System.out.println("java WordGame [filename]");
			System.exit(0);
			}
		
		// get the file contents and close immediately
		FileReader fileReader = new FileReader(args[0]);
		
		//
		// build the graph
		//
		
		// add all the vertices
		Graph graph = new Graph();
		ArrayList<String> words = new ArrayList<String>();
		while(true) {
			
			// read the next word
			String word = readWord(fileReader);
			
			// break out of the loop if the end of the file has been reached
			if (word == "eof") {
				break;
			}
			
			// get the vertex from word
			Vertex vertex = new Vertex(word);
			words.add(word);
			graph.addVertex(vertex);
			int weight;
			
			// iterate through all the other words
			for(int n = 0; n < words.size(); n++){
				weight = findWeight(graph.getVertex(words.get(n)), vertex);
				
				// add edge if valid
				if (weight == 1 || weight == 5){
					graph.getVertex(words.get(n)).connect(vertex, weight);
					vertex.connect(graph.getVertex(words.get(n)), weight);
				}
			}
		}
		
		// close the file
		fileReader.close();
		
		// items we will use in the main program loop
		HashMap<String,Vertex> vertices = graph.getVertexMap();
		Console console = System.console();
		
		// main program loop
		while(true){
			String input = console.readLine("Enter a five-letter word: ");
			
			input = input.toUpperCase();
			
			// try to pull the vertex from the graph
			Vertex vertex = graph.getVertex(input);
			if(vertex == null){
				System.out.println("The word doesn't exist in the graph, please try again.");
				continue;
			}
			
			// find the result by looking at that vertex's adjacency list
			System.out.println("The neighbors of " + input + " are :");
			int counter = 1;

			// search the adjacency list
			HashMap<Vertex, Integer> adjList = vertex.getAdjList();
			Vertex[] adjVerts = adjList.keySet().toArray(new Vertex[0]);
			for(int n = 0; n < adjVerts.length; n++){
				
				Vertex otherVertex = adjVerts[n];
				int weight = adjList.get(otherVertex);
				
				String end = "";
				if(counter == 6){
					end = "\n";
					counter = 0;
				}
				
				System.out.print("\t" + otherVertex.getWord() + " (" + String.valueOf(weight) + ")" + end);
				counter += 1;
			}
			
			System.out.println();
		}
	}
}
